src/Interpreters/SyntaxAnalyzer.cpp

```diff
@@ -346,6 +346,98 @@ void optimizeGroupBy(ASTSelectQuery * select_query, const NameSet & source_colum
         appendUnusedGroupByColumn(select_query, source_columns);
 }
 
+/// recursive traversal and check for optimizeGroupByFunctionKeys
+int optimizeGroupByFunctionKeysChecks(const ASTPtr& child, std::unordered_map<String, int> * checks)
+{
+    /// check if function has no arguments
+    if ((child->children).empty())
+    {
+        return 1;
+    }
+    for (auto & child_key : child->children)
+    {
+        if (!child_key->as<ASTFunction>())
+        {
+            if ((*checks).find(child_key->getColumnName()) == (*checks).end())
+            {
+                /// if variable of a function is not in GROUP BY keys, this function should not be deleted
+                return 1;
+            }
+        }
+        else
+        {
+            /// if function is not in keys, check for its variables
+            if ((*checks).find(child_key->getColumnName()) == (*checks).end())
+            {
+                return optimizeGroupByFunctionKeysChecks(child_key->as<ASTFunction>()->arguments, checks);
+            }
+        }
+    }
+    return 0;
+}
+
+///eliminate functions of other GROUP BY keys
+void optimizeGroupByFunctionKeys(ASTSelectQuery * select_query)
+{
+    if (!select_query->groupBy())
+    {
+        return;
+    }
+
+    auto & group_keys = select_query->groupBy()->children;
+    auto grp_by = select_query->groupBy();
+
+    ///check if optimization is not needed
+    int opt = 0;
+    for (auto & group_key : group_keys)
+    {
+        if (group_key->as<ASTFunction>())
+        {
+            opt = 1;
+            break;
+        }
+    }
+    if (opt == 0)
+    {
+        return;
+    }
+
+    ASTs modified; ///result
+    std::unordered_map<String, int> checks; ///map of pairs <structure of key (variable/function), int>
+
+    ///filling map with pairs <key name, 1>
+    for (auto & group_key : group_keys)
+    {
+        checks[group_key->getColumnName()] = 1;
+    }
+    size_t modified_size = group_keys.size();
+    for (auto & group_key : group_keys)
+    {
+        if (group_key->as<ASTFunction>())
+        {
+            if (optimizeGroupByFunctionKeysChecks(group_key->as<ASTFunction>()->arguments, &checks) == 0)
+            {
+                ///if function of a key should be deleted, mark it as 0 in checks
+                checks[group_key->getColumnName()] = 0;
+                --modified_size;
```

Comment:
Too complex logic just for a `vector::reserve()`. It would be OK to `reserve(group_keys.size())`.